test: improve upgrade E2E test traceability

jsilvela opened the pull request:
Enhanced logging and traceability in the end-to-end upgrade tests.

Closes #5331

jsilvela committed:
test: improve traceability of upgrade E2E test

Signed-off-by: Jaime Silvela <jaime.silvela@enterprisedb.com>

jsilvela commented:
/test

gabriele-wolfox commented on internal/cmd/manager/controller/controller.go:
```diff
@@ -102,9 +103,18 @@ func RunController(
 ) error {
 	ctx := context.Background()
 
+	var operatorHash string
```
I would replace lines from 106 to 112 with

```
	operatorHash, err := executablehash.Get()
	if err != nil {
		setupLog.Error(err, "could not get executable hash")
		return err
	}
```

jsilvela replied:
👍

jsilvela replied:
I would not `return err`. IMO this is a "best effort" thing.

gabriele-wolfox replied:
If we have an error here, something really bad is happening and I would exit.
But yes, I would say return or not is debatable :)

gabriele-wolfox commented on tests/e2e/upgrade_test.go:
```diff
@@ -385,6 +418,16 @@ var _ = Describe("Upgrade", Label(tests.LabelUpgrade, tests.LabelNoOpenshift), O
+			Expect(err).NotTo(HaveOccurred())
+			lines := strings.Split(lgs, "\n")
+			Expect(lines).ShouldNot(BeEmpty())
+			GinkgoWriter.Println(lines[0])
```
I would add here a caption of what we are printing.
Something like
`GinkgoWriter.Println("pod logs", lines[0])`

jsilvela replied:
Agree

gabriele-wolfox commented on tests/e2e/upgrade_test.go:
```diff
@@ -385,6 +418,16 @@ var _ = Describe("Upgrade", Label(tests.LabelUpgrade, tests.LabelNoOpenshift), O
+			Expect(err).NotTo(HaveOccurred())
+			lines := strings.Split(lgs, "\n")
+			Expect(lines).ShouldNot(BeEmpty())
+			GinkgoWriter.Println(lines[0])
```
I would add here a caption of what we are printing.
Something like
`GinkgoWriter.Println("pod logs", lines[0])`

gabriele-wolfox replied:
```suggestion
			GinkgoWriter.Println("pod logs", lines[0])
```

gabriele-wolfox commented on internal/cmd/manager/controller/controller.go:
```diff
@@ -102,9 +103,18 @@ func RunController(
 ) error {
 	ctx := context.Background()
 
+	var operatorHash string
```
I would replace lines from 106 to 112 with

```
	operatorHash, err := executablehash.Get()
	if err != nil {
		setupLog.Error(err, "could not get executable hash")
		return err
	}
```

gabriele-wolfox commented on internal/cmd/manager/controller/controller.go:
```diff
@@ -102,9 +103,18 @@ func RunController(
 ) error {
 	ctx := context.Background()
 
+	var operatorHash string
+	h, err := executablehash.Get()
+	if err != nil {
+		setupLog.Error(err, "could not get executable hash")
+	} else {
+		operatorHash = h
+	}
```
```suggestion
	operatorHash, err := executablehash.Get()
	if err != nil {
		setupLog.Error(err, "could not get executable hash")
		return err
	}
```

jsilvela committed:
Update tests/e2e/upgrade_test.go

Co-authored-by: Gabriele Quaresima <gabriele.quaresima@enterprisedb.com>
Signed-off-by: Jaime Silvela <jaime.silvela@enterprisedb.com>

jsilvela committed:
Update internal/cmd/manager/controller/controller.go

Co-authored-by: Gabriele Quaresima <gabriele.quaresima@enterprisedb.com>
Signed-off-by: Jaime Silvela <jaime.silvela@enterprisedb.com>

gabriele-wolfox commented:
/test

jsilvela commented:
/test

YanniHu1996 commented on tests/e2e/upgrade_test.go:
```diff
@@ -282,12 +305,17 @@ var _ = Describe("Upgrade", Label(tests.LabelUpgrade, tests.LabelNoOpenshift), O
 
-			var count int
+			var count, nonUpgradeCount int
+			By("checking cluster events for manager")
```
Hi @jsilvela ,  the message `By("checking cluster events for manager")` is inside the `retry.OnError`. It may be printed multiple times. Can it be moved out of the `retry.OnError`?

jsilvela replied:
In my opinion, no, but not a strong feeling.

mnencia reviewed (changes requested):
This patch combines both code and test changes. I've provided specific comments regarding the code modifications, as they alter the operator's behavior and therefore cannot be accepted as part of a test patch.

jsilvela commented on pkg/utils/discovery.go:
```diff
@@ -64,8 +64,6 @@ func (arch *AvailableArchitecture) GetHash() string {
 	}
-	arch.mx.Lock()
-	defer arch.mx.Unlock()
 	return arch.hash
```
Hmm, so we want read protection, ok

mnencia replied:
The issue is that writing on a sting variable is not guaranteed to be atomic (even if it is in many architectures). Since this may be modified in another thread, we must protect the read operation.

jsilvela replied:
then fixing so `calculateHash` returns a value.

jsilvela commented on internal/cmd/manager/controller/controller.go:
```diff
@@ -102,9 +103,16 @@ func RunController(
 ) error {
 	ctx := context.Background()
 
+	operatorHash, err := executablehash.Get()
```
do you remember why?
Just thinking, for the "smart" upgrade test we want, checking the hash in the pods is the same as the new operator, seems like a very solid test...

jsilvela commented on internal/controller/cluster_upgrade.go:
```diff
@@ -697,39 +697,51 @@ func (r *ClusterReconciler) upgradeInstanceManager(
-					return err
-				}
-			}
+		if instanceManagerIsUpgrading || instanceManagerHash == "" || instanceManagerHash == operatorHash {
+			message := fmt.Sprintf("Instance manager will skip upgrade on %s (upgrading: %t) "+
+				"(operator hash: %s — instance manager hash: %s)",
+				postgresqlStatus.Pod.Name,
+				instanceManagerIsUpgrading,
+				operatorHash[:6],
+				instanceManagerHash[:6])
+
+			r.Recorder.Event(cluster, "Normal", "SkippedInstanceManagerUpgrade", message)
+			contextLogger.Info(message)
+			continue
+		}
```
Maybe I can try to reduce. But, it's very difficult to understand what's going on with online upgrades without a bit of logging like this one.

mnencia commented:
This issue is related to https://github.com/cloudnative-pg/cloudnative-pg/issues/4423 for the hash calculation.

mnencia commented:
> - Adds the executable hash on the operator binary and displays in
the init log
> - Computes the executable hash from the instances before and after
an upgrade
> - Add events to explain why an upgrade may have been skipped

jsilvela committed:
chore: review, repurpose

Signed-off-by: Jaime Silvela <jaime.silvela@enterprisedb.com>

jsilvela commented:
/test feature_type=upgrade test_level=4

jsilvela commented:
There. That's better:
![Screenshot 2024-08-22 at 13 22 09](https://github.com/user-attachments/assets/284542c3-efbf-4df2-9de9-8e604c360213)

Asking, once again, for review.
Addressed your comments @mnencia , and also applied the changes discussed in https://github.com/cloudnative-pg/cloudnative-pg/issues/5331

jsilvela commented:
If I may say so myself ... nice

![Screenshot 2024-08-22 at 13 33 03](https://github.com/user-attachments/assets/d8f61a3b-b1be-411a-bab2-e203cc25e8ea)

mnencia committed:
chore: review of some messages

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia commented:
The code is okay, but this pull request does too many things simultaneously. I'm going to split it into multiple PRs.

mnencia committed:
chore: remove code moved to other PRs

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

sxd committed:
chore: review

Signed-off-by: Jonathan Gonzalez V. <jonathan.gonzalez@enterprisedb.com>

NiccoloFei committed:
chore: fix typo and add comments

Signed-off-by: Niccolò Fei <niccolo.fei@enterprisedb.com>

NiccoloFei commented:
/test feature_type=upgrade test_level=4

NiccoloFei committed:
fix: RetrievePgStatusFromInstance()

Signed-off-by: Niccolò Fei <niccolo.fei@enterprisedb.com>

NiccoloFei commented:
/test feature_type=upgrade test_level=4

mnencia committed:
Merge branch 'main' into dev/upgrade-tests

mnencia committed:
Merge branch 'main' into dev/upgrade-tests
